<launch>

    <arg name="use_sim_time" default="false"/>
    
    <group>
        <include file="$(find-pkg-share pingpongbot_description)/launch/view_robot.launch.xml" >
            <arg name="use_jsp" value="none"/>
        </include>
    </group>

    <node pkg="pingpongbot_control" exec="jointStateUpdate">
        <remap from="wheel_speeds" to="wheel_speeds"/>
        <remap from="wheel_angles" to="wheel_angles"/>
        <remap from="joint_states" to="joint_states"/>
    </node>

    <node pkg="pingpongbot_control" exec="odomUpdate">
        <remap from="wheel_speeds" to="wheel_speeds"/>
        <remap from="wheel_angles" to="wheel_angles"/>
        <remap from="joint_states" to="joint_states"/>
        <remap from="cmd_vel" to="cmd_vel"/>
        <remap from="reached_goal" to="reached_goal"/>
        <remap from="odom" to="odom"/>
        <remap from="odometry/filtered" to="odometry/filtered"/>
        <param from="$(find-pkg-share pingpongbot_control)/config/robot_dimensions.yaml"/>
    </node>

    <node pkg="pingpongbot_control" exec="controller">
        <remap from="cmd_vel" to="cmd_vel"/>
        <remap from="goal_pose" to="goal_pose"/>
        <remap from="heartbeat" to="heartbeat"/>
        <remap from="error" to="error"/>
        <remap from="reached_goal" to="reached_goal"/>
        <param from="$(find-pkg-share pingpongbot_control)/config/controller.yaml"/>
    </node>

    <node pkg="robot_localization" type="ekf_node" name="ekf_filter_node" output="screen">
        <param name="use_sim_time" value="$(arg use_sim_time)"/>
        <param name="odom_frame" value="odom"/>
        <param name="base_link_frame" value="base_footprint"/>
        <param name="world_frame" value="odom"/>
        <param name="imu0" value="imu"/>
        <param name="odom0" value="odom"/>
        <param name="imu0_config" value="$(find-pkg-share pingpongbot_control)/config/ekf.yaml"/>
        <remap from="imu" to="imu"/>
        <remap from="odom" to="odom"/>
        <remap from="odometry/filtered" to="odometry/filtered"/>
    </node>

    <node pkg="rqt_reconfigure" exec="rqt_reconfigure"/>

    <!-- <node pkg="rqt_plot" exec="rqt_plot"/> -->

</launch>